R/validate-policy.R

```diff
@@ -1,26 +1,35 @@
 #' Validate Policy File
 #' 
 #' Flattens and validates policy file
-#'  @param policy YAML policy file
-#'  @param gs Gradescope data
-#'  @importFrom purrr map discard
-#'  @export
+#' @param policy YAML policy file
+#' @param gs Gradescope data
+#' @importFrom purrr map discard
+#' @export
 validate_policy <- function(policy, gs){
```

Comment:
How often do you anticipate calling `validate_policy`? After each call to `calculate_grades`? If so, speed, while not the only issue, is one consideration to keep in mind due to the delay this will cause in the app.

One thing this bring up is modularity of `calculate_grades()`. The way that it's currently written, can `policy` be a subtree of the full policy file? If that is possible, that would open up a more efficient way to recalculate grades when a user edits one category on the policy file - it'd just recalculate for that small subtree (and the uptree dependencies).

If we have that property in `calculate_grades`, seems like we'd be able to echo it here.

Reply:
I was under the impression that validate_policy would be run as an observeEvent whenever policy$categories changes; I believe calculate_grades is already run as an observeEvent in a similar fashion, and validate_policy needs to be run before calculate_grades in order to add defaults. Subtree approach is possible, I think, because it is basically a mini-policy itself. So yes, if we have that subtree property, validate_policy should be able to support it

Reply:
As far as the usage of validate_policy, I am not sure exactly where and when it would be used as I am not an architect of those parts of the code. For your modularity thoughts, I considered a solution that tracks dependencies for validate_policy, which gave me some insight into how that could be done. While I think it is possible to do so, the policy file's lack of a proper tree structure does add challenge. For the purpose of modularity during grade calculation, perhaps we could also have the tree "upside down"--with assignments as roots and categories as children--to only update certain portions of the grading. Also, to make each category identically a policy of its own, we might want to remove the coursewide info and the "categories" tag from the top level of the nesting once the policy is loaded into R in order to make each category truly like a policy file -- or here perhaps the policy file more like a category. Right now, they are not fully symmetrical data structures.